editor/src/clj/editor/color_dropper.clj

```diff
@@ -0,0 +1,137 @@
+      diameter (* pixel-size (count pixel-seq))
+      radius ^double (/ diameter 2)]
+  (defn- mouse-move-handler!
+    [view-node ^Canvas canvas ^MouseEvent e]
+    (.consume e)
+    (when-let [image ^WritableImage (g/node-value view-node :image)]
+      (let [graphics-context ^GraphicsContext (.getGraphicsContext2D canvas)
+            pixel-reader ^PixelReader (.getPixelReader image)
+            mouse-x (.getX e)
+            mouse-y (.getY e)]
+        (when (in-bounds? image mouse-x mouse-y)
+          (g/set-property! view-node :color (.getColor pixel-reader mouse-x mouse-y)))
+        (.clearRect graphics-context (- mouse-x radius) (- mouse-y radius) diameter diameter)
+        (.setClip canvas (create-mask! mouse-x mouse-y radius))
+        (.setStroke graphics-context Color/GRAY)
+
+        (doseq [^int x-step pixel-seq
+                ^int y-step pixel-seq
+                :let [x (+ mouse-x (* x-step pixel-size) (- pixel-center))
+                      y (+ mouse-y (* y-step pixel-size) (- pixel-center))]]
+          (when (in-bounds? image x y)
+            (->> (.getColor pixel-reader (+ mouse-x x-step) (+ mouse-y y-step))
+                 (paint-pixel! graphics-context x y pixel-size))))
+
+        (doto graphics-context
+          (.strokeOval (- mouse-x radius) (- mouse-y radius) diameter diameter)
+          (.setStroke Color/RED)
+          (.strokeRect (- mouse-x pixel-center) (- mouse-y pixel-center) pixel-size pixel-size))))))
+
+(defn- remove-dropper-area!
+  [view-node]
+  (let [dropper-area ^StackPane (g/node-value view-node :dropper-area)
+        root ^StackPane (ui/main-root)]
+    (g/set-property! view-node :dropper-area nil)
+    (-> (.getChildren root)
+        (.remove dropper-area))))
+
+(defn- apply-and-deactivate!
+  [view-node pick-fn]
+  (remove-dropper-area! view-node)
+  (pick-fn (g/node-value view-node :color)))
+
+(defn- key-pressed-handler!
+  [view-node pick-fn ^KeyEvent event]
+  (condp = (.getCode event)
+    KeyCode/ENTER (apply-and-deactivate! view-node pick-fn)
+    KeyCode/ESCAPE (remove-dropper-area! view-node)
+    nil))
+
+(defn make-color-dropper! [graph]
+  (g/make-node! graph ColorDropper))
+
+(defn activate!
+  [view-node pick-fn]
+  (let [main-view ^StackPane (ui/main-root)
+        canvas (Canvas. (.getWidth main-view) (.getHeight main-view))
+        dropper-area (doto (StackPane.)
+                       (.setCursor Cursor/NONE)
+                       (ui/add-child! canvas)
+                       (.setStyle "-fx-background-color: transparent;"))]
+    (ui/add-child! main-view dropper-area)
+    (g/set-property! view-node :dropper-area dropper-area)
+
+    (.bind (.widthProperty canvas) (.widthProperty dropper-area))
+    (.bind (.heightProperty canvas) (.heightProperty dropper-area))
+
+    (ui/observe (.widthProperty main-view) (fn [_ _ _] (capture! view-node canvas)))
+    (ui/observe (.heightProperty main-view) (fn [_ _ _] (capture! view-node canvas)))
```

Comment:
Don't we need to un-observe these when we dismiss the view? Otherwise we'll keep adding new observers to the width and height properties of the `main-view`, right?

Reply:
Correct. I tried observing canvas initially but it didn't work, and then forgot to remove the listener from main.